### PR TITLE
Feature/issue 756 rename cov_exp_quad -> gp_exp_quad_cov

### PR DESCRIPTION
**Pull request by drezap**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Deprecate `cov_exp_quad.cpp`, retaining back compitability and test cases.
#### Intended Effect:
To rename 
#### How to Verify:
All old test cases work.
#### Side Effects:
None.
#### Copyright and Licensing

Copyright 2018 Andre Zapico

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by drezap** [comment]
Hi -

I believe the only function that needed to be renamed was `cov_exp_quad`, i.e. `gp_dot_prod_cov`. Can someone review this so I can close this pull request? 

Thanks!

**Comment by drezap** [comment]
closing PR until finished

File: stan/math/prim/mat/fun/dot_product.hpp [hunk 1]
```diff
@@ -19,8 +19,8 @@ namespace math {
  * size or if they are both not vector dimensioned.
  */
 template <int R1, int C1, int R2, int C2>
-inline double dot_product(const Eigen::Matrix<double, R1, C1>& v1,
-                          const Eigen::Matrix<double, R2, C2>& v2) {
+inline double dot_product(const Eigen::Matrix<double, R1, C1> &v1,
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/dot_product.hpp]: Am I missing something or are the changes to the dot_product.hpp just changed positions of & and * and an added #include that I am not sure if it is needed.
I am guessing this file was changed severly in the process but was then reverted back? Can you revert these changes if they have no effect?

File: test/unit/math/prim/mat/fun/dot_product_test.cpp [hunk 2]
```diff
@@ -1,2 +1,2 @@
-#include <stan/math/prim/mat.hpp>
 #include <gtest/gtest.h>
+#include <stan/math/prim/mat.hpp>
```
Comment (rok-cesnovar) [inline test/unit/math/prim/mat/fun/dot_product_test.cpp]: Same as for the dot_product.hpp I guess. Please revert this change back if it has no other effect.

File: stan/math/rev/mat/fun/cov_exp_quad.hpp [hunk 3]
```diff
@@ -207,35 +217,14 @@ inline typename boost::enable_if_c<
...
-  cov.coeffRef(x_size - 1, x_size - 1).vi_ = baseVari->cov_diag_[x_size - 1];
-  return cov;
+  return gp_exp_quad_cov(x, sigma, l);
 }
 
 /**
  * Returns a squared exponential kernel.
  *
+ * @deprecated use <code>gp_exp_quad_cov_vari</code>
+ *
```
Comment (rok-cesnovar) [inline stan/math/rev/mat/fun/cov_exp_quad.hpp]: No need for this empty line here.

File: stan/math/prim/mat/fun/cov_exp_quad.hpp [hunk 4]
```diff
@@ -18,6 +19,8 @@ namespace math {
 /**
  * Returns a squared exponential kernel.
  *
+ * @deprecated use <code>gp_exp_quad_cov</code>
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/cov_exp_quad.hpp]: Function with @deprecated usually have other comments removed (no explanation of what the function does, no @tparam, etc). See https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/prob/dirichlet_log.hpp#L37

File: stan/math/prim/mat/fun/cov_exp_quad.hpp [hunk 5]
```diff
@@ -36,38 +39,14 @@ inline
...
-  }
-  cov(x_size - 1, x_size - 1) = sigma_sq;
-  return cov;
+  return gp_exp_quad_cov(x, sigma, length_scale);
 }
 
 /**
  * Returns a squared exponential kernel.
  *
+ * @deprecated use <code>gp_exp_quad_cov</code>
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/cov_exp_quad.hpp]: Remove comments other than the @depracted line. See above.

File: stan/math/prim/mat/fun/cov_exp_quad.hpp [hunk 6]
```diff
@@ -86,47 +65,14 @@ inline
...
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/cov_exp_quad.hpp]: Remove comments other than the @depracted line. See above.

File: stan/math/prim/mat/fun/cov_exp_quad.hpp [hunk 7]
```diff
@@ -146,35 +92,14 @@ inline typename Eigen::Matrix<
...
-    }
-  }
-  return cov;
+  return gp_exp_quad_cov(x1, x2, sigma, length_scale);
 }
 
 /**
  * Returns a squared exponential kernel.
  *
+ * @deprecated use <code>gp_exp_quad_cov</code>
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/cov_exp_quad.hpp]: Remove comments other than the @depracted line. See above.

File: stan/math/prim/mat/fun/gp_exp_quad_cov.hpp [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+namespace stan {
+namespace math {
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * @tparam T_x type of std::vector of elements
+ * @tparam T_sigma type of sigma
+ * @tparam T_l type of length scale
+ *
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/gp_exp_quad_cov.hpp]: No need for this empty line.

File: stan/math/prim/mat/fun/gp_exp_quad_cov.hpp [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+  return cov;
+}
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * @tparam T_x type of std::vector of elements
+ * @tparam T_sigma type of sigma
+ * @tparam T_l type of std::vector of length scale
+ *
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/gp_exp_quad_cov.hpp]: No need for the empty line.

File: stan/math/prim/mat/fun/gp_exp_quad_cov.hpp [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * @tparam T_x1 type of first std::vector of elements
+ * @tparam T_x2 type of second std::vector of elements
+ * @tparam T_sigma type of sigma
+ * @tparam T_l type of of length scale
+ *
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/gp_exp_quad_cov.hpp]: Empty line, see above.

File: stan/math/prim/mat/fun/gp_exp_quad_cov.hpp [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * @tparam T_x1 type of first std::vector of elements
+ * @tparam T_x2 type of second std::vector of elements
+ * @tparam T_sigma type of sigma
+ * @tparam T_l type of length scale
+ *
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/gp_exp_quad_cov.hpp]: Empty line, see above.

File: stan/math/rev/mat/fun/cov_exp_quad.hpp [hunk 12]
```diff
@@ -28,6 +28,8 @@ namespace math {
  * matrix, along with a pointer to the vari for sigma,
  * and the vari for l.
  *
+ * @deprecated use <code>gp_exp_quad_cov_vari</code>
```
Comment (rok-cesnovar) [inline stan/math/rev/mat/fun/cov_exp_quad.hpp]: Remove comments other than the @depracted line. See above.

File: stan/math/rev/mat/fun/cov_exp_quad.hpp [hunk 13]
```diff
@@ -59,6 +61,8 @@ class cov_exp_quad_vari : public vari {
    * controlled to the second argument to
    * vari's constructor.
    *
+   * @deprecated use <code>gp_exp_quad_cov_vari</code>
```
Comment (rok-cesnovar) [inline stan/math/rev/mat/fun/cov_exp_quad.hpp]: Remove comments other than the @depracted line. See above.

File: stan/math/rev/mat/fun/cov_exp_quad.hpp [hunk 14]
```diff
@@ -120,6 +124,8 @@ class cov_exp_quad_vari : public vari {
  * matrix, along with a pointer to the vari for sigma,
  * and the vari for l.
  *
+ * @deprecated use <code>gp_exp_quad_cov_vari</code>
```
Comment (rok-cesnovar) [inline stan/math/rev/mat/fun/cov_exp_quad.hpp]: Remove comments other than the @depracted line. See above.

File: stan/math/rev/mat/fun/cov_exp_quad.hpp [hunk 15]
```diff
@@ -149,6 +155,8 @@ class cov_exp_quad_vari<T_x, double, T_l> : public vari {
    * controlled to the second argument to
    * vari's constructor.
    *
+   * @deprecated use <code>gp_exp_quad_cov_vari</code>
```
Comment (rok-cesnovar) [inline stan/math/rev/mat/fun/cov_exp_quad.hpp]: Remove comments other than the @depracted line. See above.

File: stan/math/rev/mat/fun/cov_exp_quad.hpp [hunk 16]
```diff
@@ -194,6 +202,8 @@ class cov_exp_quad_vari<T_x, double, T_l> : public vari {
 /**
  * Returns a squared exponential kernel.
  *
+ * @deprecated use <code>gp_exp_quad_cov_vari</code>
```
Comment (rok-cesnovar) [inline stan/math/rev/mat/fun/cov_exp_quad.hpp]: Remove comments other than the @depracted line. See above.

File: stan/math/rev/mat/fun/cov_exp_quad.hpp [hunk 17]
```diff
@@ -207,35 +217,14 @@ inline typename boost::enable_if_c<
...
-  }
-  cov.coeffRef(x_size - 1, x_size - 1).vi_ = baseVari->cov_diag_[x_size - 1];
-  return cov;
+  return gp_exp_quad_cov(x, sigma, l);
 }
 
 /**
  * Returns a squared exponential kernel.
  *
+ * @deprecated use <code>gp_exp_quad_cov_vari</code>
```
Comment (rok-cesnovar) [inline stan/math/rev/mat/fun/cov_exp_quad.hpp]: Remove comments other than the @depracted line. See above.

**Comment by rok-cesnovar** [comment]
Please also take a look at the Jenkins errors:http://d1m1s1b1.stat.columbia.edu:8080/blue/organizations/jenkins/Math%20Pipeline/detail/PR-772/1/pipeline

File: stan/math/prim/mat/fun/dot_product.hpp (hunk 1 above)
Comment (drezap) [reply stan/math/prim/mat/fun/dot_product.hpp]: @rok-cesnovar I believe this was some autoformatting thing, stuff like this is handled automatically...

File: stan/math/rev/mat/fun/cov_exp_quad.hpp (hunk 14 above)
Comment (drezap) [reply stan/math/rev/mat/fun/cov_exp_quad.hpp]: feel free to just say "same goes for the other files" or something next time...

**Comment by drezap** [comment]
@rok-cesnovar thanks for the review! Updating now!

**Comment by rok-cesnovar** [comment]
Thanks for the quick response. 
For the jenkins errors I think adding instance() to 
`cov_diag_(ChainableStack::memalloc_.alloc_array<vari *>(size_)`
so that you get 
`cov_diag_(ChainableStack::instance().memalloc_.alloc_array<vari *>(size_)`

 would clear most if not all of the errors.

**Comment by rok-cesnovar** [comment]
The test-headers now pass so this is close to done.
Would you mind reverting `stan/math/prim/mat/fun/dot_product.hpp` and `test/unit/math/prim/mat/fun/dot_product_test.cpp` to the previous state since these changes are irrelevant? I can help you out with those, if I can access the branch.

When those are reverted and all the Jenkins stuff passes I think we can merge.

**Comment by drezap** [comment]
@rok-cesnovar updated the headers for dot_product and dot_product_test, thanks

**Comment by drezap** [comment]
The change in the location of the `&` is just autoformatting. Dot product was implemented before autoformatting was in place, and then now that we are using autoformatting, this changes the location. Is this ok? If I go and revert it, the autoformatting is going to change it all back again...

**Comment by rok-cesnovar** [comment]
Ah, you are right. Forgot about that. Sorry, this is my first Stan Math PR review. I will get the hang of this eventually :)

**Comment by rok-cesnovar** [comment]
@syclik I think this is ready to merge. Should I merge this or how does this go from here? First time doing this, so just wanted a conformation.

**Comment by syclik** [comment]
@rok-cesnovar, merge! If you think there's something someone else needs to look at, please ask for it, but otherwise, merge away.
